Improve Ubuntu Advantage module's idempotency

chifac08 opened the pull request:
## Proposed Commit Message

```
Ubuntu Advantage: Improved idempotency

Ubuntu Advantage module is now capable of detecting already
activated subscriptions or/and enabled services. Furthermore,
when "enable" parameter is set, no "default" services
will be enabled anymore.
```

## Additional Context
If `instance-id` changes due to updated user or meta-data, cloud-init
will run again and UA causes an error when a machine has an already
attached subscription and/or service.

Ubuntu Advantage is now capable of ignoring the following messages:

```
Stderr: This machine is already attached to 'X'
        To use a different subscription first run: sudo ua detach.

Stdout: One moment, checking your subscription first
        <Service> is already enabled.
        See: sudo ua status
```
        
The module won't set cloud-init status to error when 
detecting those messages. Nevertheless, it will print error
messages in cloud-init log for easier troubleshooting.

Furthermore, when "enable" parameter is set, the module
only activates the list of services given rather than 
activating default services + list of enabled.

## Test Steps
Tested with the following user-data:

1. user-data with no "enable":
```
ubuntu_advantage:
  token: "{{ ds.meta_data.ua_token }}"
```

3. cloud-init clean --logs --reboot

4. cloud-init status should not state "error" and "default services" should be enabled:
```
root@test:/home/ubuntu# cloud-init status -l
status: done
time: Fri, 19 Aug 2022 06:49:50 +0000
detail:
DataSourceNoCloud [seed=/var/lib/cloud/seed/nocloud,/dev/sdc][dsmode=net]

SERVICE          ENTITLED  STATUS    DESCRIPTION
cc-eal           yes       n/a       Common Criteria EAL2 Provisioning Packages
esm-infra        yes       enabled   UA Infra: Extended Security Maintenance (ESM)
fips             yes       disabled  NIST-certified core packages
fips-updates     yes       disabled  NIST-certified core packages with priority security updates
livepatch        yes       enabled   Canonical Livepatch service
usg              yes       disabled  Security compliance and audit tools
```

5. ua detach

6. user-data with "enable":
```
ubuntu_advantage:
  token: "{{ ds.meta_data.ua_token }}"
  enable:
    - livepatch
```

7. cloud-init clean --logs --reboot

8. cloud-init status should not state "error" and only "livepatch" service should be enabled:
```
root@test:/home/ubuntu# cloud-init status -l
status: done
time: Fri, 19 Aug 2022 06:49:50 +0000
detail:
DataSourceNoCloud [seed=/var/lib/cloud/seed/nocloud,/dev/sdc][dsmode=net]
```

## Checklist:
 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/contributing.html)
 - [x] I have updated or added any unit tests accordingly
 - [x] I have updated or added any documentation accordingly

aciba90 reviewed (changes requested):
Thanks, @chifac08, for improving cloud-init.

I think we should adapt the module doc to reflect this behavior. We changed it to reflect the previous behavior in https://github.com/canonical/cloud-init/pull/1487.

aciba90 commented on cloudinit/config/cc_ubuntu_advantage.py:
```diff
@@ -208,20 +212,36 @@ def configure_ua(token=None, enable=None, config=None):
+            msg = 'Service "{service}" already enabled:\n{warning}'.format(
+                service=service, warning=str(warning)
+            )
+            util.logexc(LOG, msg)
```
Shouldn't be this logged as a warning?

```suggestion
            LOG.warning(msg)
```

chifac08 commented:
> Thanks, @chifac08, for improving cloud-init.
> 
> I think we should adapt the module doc to reflect this behavior. We changed it to reflect the previous behavior in #1487.

@aciba90 Thanks for your quick response. Adjusted documentation regarding the 'enable' parameter. Do you also want me to document the new behavior that "already attached subscriptions and/or enabled services" are logged as warning rather than causing cloud-init to fail? I am not sure if we should document it because the new behavior, introduced with this PR, assures idempotency which should be seen as default. Let me know what you think :)

Furthermore, I implemented source code changes you requested/suggested.

aciba90 commented on cloudinit/config/cc_ubuntu_advantage.py:
```diff
@@ -208,20 +213,36 @@ def configure_ua(token=None, enable=None, config=None):
         util.logexc(LOG, msg)
-        raise RuntimeError(msg) from e
+
+        if not re.search("This machine is already attached to", str(e)):
```
Instead of relaying on the error message, we could allow the status code `2` in `subp.subp`. That have the added benefit of not logging an error message in this case, which could confuse the user.

See the main message, containing a diff proposal.

aciba90 commented on cloudinit/config/cc_ubuntu_advantage.py:
```diff
@@ -208,20 +213,36 @@ def configure_ua(token=None, enable=None, config=None):
+            raise RuntimeError(msg) from e
+
     enable_errors = []
+    enable_warnings = []
```
Thinking more about this, wouldn't be better to just log this with DEBUG level? We are handling this case and there is nothing to justify it as WARNING, just to improve debug-ability.

See the main message, containing a diff proposal.

aciba90 reviewed (changes requested):
> Do you also want me to document the new behavior that "already attached subscriptions and/or enabled services" are logged as warning rather than causing cloud-init to fail?

I think it is enough as-is. I see that as internal/developer documentation more than user-facing.

Here is a diff proposing two improvements, 

1. Not log errors when the system is already ua-attached.
2. Logs about already enabled services as DEBUG.

```diff
diff --git a/cloudinit/config/cc_ubuntu_advantage.py b/cloudinit/config/cc_ubuntu_advantage.py
index a3bcde1c..ce7f80e7 100644
--- a/cloudinit/config/cc_ubuntu_advantage.py
+++ b/cloudinit/config/cc_ubuntu_advantage.py
@@ -207,35 +207,26 @@ def configure_ua(token=None, enable=None, config=None):
         attach_cmd = ["ua", "attach", token]
     LOG.debug("Attaching to Ubuntu Advantage. %s", " ".join(attach_cmd))
     try:
-        subp.subp(attach_cmd)
+        # Allow `ua attach` to fail in already attached machines
+        subp.subp(attach_cmd, rcs={0, 2})
     except subp.ProcessExecutionError as e:
         msg = "Failure attaching Ubuntu Advantage:\n{error}".format(
             error=str(e)
         )
         util.logexc(LOG, msg)
-
-        if not re.search("This machine is already attached to", str(e)):
-            raise RuntimeError(msg) from e
+        raise RuntimeError(msg) from e
 
     enable_errors = []
-    enable_warnings = []
     for service in enable:
         try:
             cmd = ["ua", "enable", "--assume-yes", service]
             subp.subp(cmd, capture=True)
         except subp.ProcessExecutionError as e:
             if re.search("is already enabled.", str(e)):
-                enable_warnings.append((service, e))
+                LOG.debug('Service "%s" already enabled.', service)
             else:
                 enable_errors.append((service, e))
 
-    if enable_warnings:
-        for service, warning in enable_warnings:
-            msg = 'Service "{service}" already enabled:\n{warning}'.format(
-                service=service, warning=str(warning)
-            )
-            LOG.warning(msg)
-
     if enable_errors:
         for service, error in enable_errors:
             msg = 'Failure enabling "{service}":\n{error}'.format(
diff --git a/tests/unittests/config/test_cc_ubuntu_advantage.py b/tests/unittests/config/test_cc_ubuntu_advantage.py
index 782a264e..fef13b1a 100644
--- a/tests/unittests/config/test_cc_ubuntu_advantage.py
+++ b/tests/unittests/config/test_cc_ubuntu_advantage.py
@@ -44,7 +44,7 @@ class TestConfigureUA:
             # When token is provided, attach the machine to ua using the token.
             pytest.param(
                 {"token": "SomeToken"},
-                [mock.call(["ua", "attach", "SomeToken"])],
+                [mock.call(["ua", "attach", "SomeToken"], rcs={0, 2})],
                 [
                     (
                         MPATH,
@@ -57,7 +57,7 @@ class TestConfigureUA:
             # When services is an empty list, do not auto-enable attach.
             pytest.param(
                 {"token": "SomeToken", "enable": []},
-                [mock.call(["ua", "attach", "SomeToken"])],
+                [mock.call(["ua", "attach", "SomeToken"], rcs={0, 2})],
                 [
                     (
                         MPATH,
@@ -72,7 +72,7 @@ class TestConfigureUA:
                 {"token": "SomeToken", "enable": ["fips"]},
                 [
                     mock.call(
-                        ["ua", "attach", "--no-auto-enable", "SomeToken"]
+                        ["ua", "attach", "--no-auto-enable", "SomeToken"], rcs={0, 2}
                     ),
                     mock.call(
                         ["ua", "enable", "--assume-yes", "fips"], capture=True
@@ -93,7 +93,7 @@ class TestConfigureUA:
                 {"token": "SomeToken", "enable": "fips"},
                 [
                     mock.call(
-                        ["ua", "attach", "--no-auto-enable", "SomeToken"]
+                        ["ua", "attach", "--no-auto-enable", "SomeToken"], rcs={0, 2}
                     ),
                     mock.call(
                         ["ua", "enable", "--assume-yes", "fips"], capture=True
@@ -118,7 +118,7 @@ class TestConfigureUA:
             # When services not string or list, warn but still attach
             pytest.param(
                 {"token": "SomeToken", "enable": {"deffo": "wont work"}},
-                [mock.call(["ua", "attach", "SomeToken"])],
+                [mock.call(["ua", "attach", "SomeToken"], rcs={0, 2})],
                 [
                     (
                         MPATH,
@@ -148,7 +148,7 @@ class TestConfigureUA:
     def test_configure_ua_attach_on_service_error(self, m_subp, caplog):
         """all services should be enabled and then any failures raised"""
 
-        def fake_subp(cmd, capture=None):
+        def fake_subp(cmd, capture=None, rcs=None):
             fail_cmds = [
                 ["ua", "enable", "--assume-yes", svc] for svc in ["esm", "cc"]
             ]
@@ -168,7 +168,7 @@ class TestConfigureUA:
         ):
             configure_ua(token="SomeToken", enable=["esm", "cc", "fips"])
         assert m_subp.call_args_list == [
-            mock.call(["ua", "attach", "--no-auto-enable", "SomeToken"]),
+            mock.call(["ua", "attach", "--no-auto-enable", "SomeToken"], rcs={0, 2}),
             mock.call(["ua", "enable", "--assume-yes", "esm"], capture=True),
             mock.call(["ua", "enable", "--assume-yes", "cc"], capture=True),
             mock.call(["ua", "enable", "--assume-yes", "fips"], capture=True),
@@ -195,7 +195,7 @@ class TestConfigureUA:
             token="SomeToken", config=["http_proxy=http://some-proxy.net:3128"]
         )
         assert [
-            mock.call(["ua", "attach", "SomeToken"])
+            mock.call(["ua", "attach", "SomeToken"], rcs={0, 2})
         ] == m_subp.call_args_list
         assert (
             MPATH,
```

Additionally, it would be nice to add unit tests covering the new functionality.

Thanks, @chifac08.

chifac08 commented:
@aciba90 Thanks for your review. I did not knew that "ua attach" returns exit code 2 when it is already attached to an subscription. That's a nice feature. Nevertheless, I implemented your requested improvements/changes. Please review again.

aciba90 reviewed (approved):
LGTM. Thanks, @chifac08.

I leave a reproducer script, just in case another reviewer wants to reproduce it:

<details>
  <summary>reproduce</summary>
  
  ```sh
  #!/bin/bash -eux

NAME=ua-idemp
BASE_DIR="<path_to_local>/cloud-init"
UA_TOKEN='<valid_token>'

cat <<EOF > /tmp/user-data
#cloud-config
ubuntu_advantage:
  token: $UA_TOKEN
  enable:
  - usg
EOF

lxc delete -f $NAME > /dev/null || true

lxc init ubuntu-daily:focal $NAME --vm -c user.user-data="$(cat /tmp/user-data)"
lxc config device add "$NAME" host-cloud-init disk source="$BASE_DIR/cloudinit" path=/usr/lib/python3/dist-packages/cloudinit
sleep 1
lxc start $NAME
sleep 30
lxc exec $NAME -- cloud-init status --wait
# lxc shell $NAME
lxc exec $NAME -- cloud-init clean --logs
lxc exec $NAME -- ua status
lxc stop $NAME

cat <<EOF > /tmp/user-data
#cloud-config
ubuntu_advantage:
  token: $UA_TOKEN
  enable:
  - usg
EOF

lxc config set $NAME user.user-data "$(cat /tmp/user-data)"
sleep 10
lxc start $NAME
sleep 30
lxc exec $NAME -- cloud-init status --wait
lxc exec $NAME -- ua status

  ```
</details>

aciba90 commented:
This PR, if accepted, could be merged before https://github.com/canonical/cloud-init/pull/1583 and I will solve possible conflicts there.

blackboxsw commented on cloudinit/config/cc_ubuntu_advantage.py:
```diff
@@ -199,29 +200,40 @@ def configure_ua(token=None, enable=None, config=None):
     enable_errors = []
     for service in enable:
         try:
             cmd = ["ua", "enable", "--assume-yes", service]
```
Can we adapt this cmd so we parse --format json output which should be more stable across languages than the CLI stdout.

```suggestion
            cmd = ["ua", "enable", "--assume-yes", service, "--format", "json"]
```

aciba90 replied:
Tracking in https://warthogs.atlassian.net/browse/SC-1253

blackboxsw commented on cloudinit/config/cc_ubuntu_advantage.py:
```diff
@@ -199,29 +200,40 @@ def configure_ua(token=None, enable=None, config=None):
             subp.subp(cmd, capture=True)
         except subp.ProcessExecutionError as e:
-            enable_errors.append((service, e))
+            if re.search("is already enabled.", str(e)):
```
If using `--format json` above we can source json output of e.stdout to determine the failure type instead of patial matching the English string.

```suggestion
            error_codes = [e.get("message_code") for e in json.loads(e.stdout)["errors"]]
            if 'service-already-enabled' in error_codes:
```

blackboxsw replied:
@chifac08 thanks so much for this contribution. My minor PR change request can be carried forward to a separate PR, I'll dismiss my comment here and we can land your work. Again thanks for patience and effort here to improve cloud-init and UA integration.

blackboxsw reviewed (commented):
Looks good to me. Although your code to treat  `enable: [a,b,c]` as an explicit list of services to enable represents a change in behavior from the way cloud-init currently works, I think the current behavior is a "bug" in that originally neither cloud-init nor UA could originally prevent the auto-enable of all contract entitled services during attach.

I agree that it makes sense to treat the behavior for the `enable: [a, b,c ]` configuration as explicit enable of UA services  instead of implicitly supplementing the services defined in an unknown contract to which we are attaching.

Minor suggestion to use `--format json` if we can during our validation of service-is-enabled message_codes.

blackboxsw reviewed (approved):
LGTM! Carrying minor error handling comment forward to a separate PR.

chifac08 commented:
> Looks good to me. Although your code to treat `enable: [a,b,c]` as an explicit list of services to enable represents a change in behavior from the way cloud-init currently works, I think the current behavior is a "bug" in that originally neither cloud-init nor UA could originally prevent the auto-enable of all contract entitled services during attach.
> 
> I agree that it makes sense to treat the behavior for the `enable: [a, b,c ]` configuration as explicit enable of UA services instead of implicitly supplementing the services defined in an unknown contract to which we are attaching.
> 
> Minor suggestion to use `--format json` if we can during our validation of service-is-enabled message_codes.

This PR fixes Bug LP: [#1824556](https://bugs.launchpad.net/cloud-init/+bug/1824556)
Do you agree?

aciba90 commented:
> > Looks good to me. Although your code to treat `enable: [a,b,c]` as an explicit list of services to enable represents a change in behavior from the way cloud-init currently works, I think the current behavior is a "bug" in that originally neither cloud-init nor UA could originally prevent the auto-enable of all contract entitled services during attach.
> > I agree that it makes sense to treat the behavior for the `enable: [a, b,c ]` configuration as explicit enable of UA services instead of implicitly supplementing the services defined in an unknown contract to which we are attaching.
> > Minor suggestion to use `--format json` if we can during our validation of service-is-enabled message_codes.
> 
> This PR fixes Bug LP: [#1824556](https://bugs.launchpad.net/cloud-init/+bug/1824556) Do you agree? If yes, do you want me to leave a comment on that bug mentioning this PR?
> 
> thx

+1. Please do so, and change the status to fix committed.
